### DISCUSSION TITLE
Unit testing code with WeakRef/FinalizationRegistry, like this LinkedAbortController

### DISCUSSION THREAD

**Discussion by noseratio** [nodejs/help#4967] [Q&A] [answered]
TLTR: is there a way to force GC in the current version of Node?

I'm looking at using `AbortController` for token-based cooperative cancellation, where it's common to have a hierarchy of linked token sources. That'd be child `AbortController` objects which would listen for `abort` event on their parent `AbortController.signal`. Here is a related WHATWG [discussion thread](https://github.com/whatwg/dom/issues/920). 

So I'm experimenting with a version of `LinkedAbortController` that uses `WeakRef` and `FinalizationRegistry` (as described in [V8's Weak references and finalizers](https://v8.dev/features/weak-references)), in attempt to help freeing the `parentAbortSignal.removeEventListener('abort', listener)` registrations for orphaned child `AbortController` instances. The cleanup should be taking place as a part of GC, when there is no more references left to a particular instance of `LinkedAbortController`, besides the weak ref from its internal `listener` for parent signal.

The [full code gist is here](https://gist.github.com/noseratio/f037cd08041a72e31e3899747cb85462).  I'm struggling to come up with a deterministic test for this in Node. Even when I call `global.gc(true)` in a loop with a delay [like this](https://gist.github.com/noseratio/f037cd08041a72e31e3899747cb85462#file-linkedabortcontroller-js-L66), I still need a final ~10s delay to see the finalizer callbacks invoked. 

Is there a more predictable way of forcing GC in Node? I'm using the latest v15.4.0. 

OTOH, I'm able to test it in the browser, the finalizers gets called when I click the Trash icon ("Collect garbage") in the Performance section of DevTools. So, for browsers, I should be able to automate this with Puppeteer.

Thanks for any help!

**Comment by jasnell** [comment, marked as answer]
Forcing? Not really. You could try the natives syntax function `%CollectGarbage` but I believe that has the same impact as calling `global.gc()`

**Reply by noseratio** [reply]
> Forcing? Not really. You could try the natives syntax function `%CollectGarbage` but I believe that has the same impact as calling `global.gc()`

@jasnell This is brilliant, thanks! Unlike `global.gc(true) `, calling `%CollectGarbage(true)` triggers the GC right away, it's exactly what I was looking for. 

To others who (like me) never heard about [built-in functions](https://v8.dev/docs/builtin-functions) before and might have ended up here, you need to run it like this to enable `%CollectGarbage(true)`:

```text
node --allow-natives-syntax LinkedAbortController.js
```

To enable `global.gc(true)`:

```text
node --expose-gc --trace-gc LinkedAbortController.js
```

**Reply by benjamingr** [reply]
See https://github.com/nodejs/node/pull/36607/files#diff-7177d219f295c5297f8e5d119cea177fc0b50c1df33161729ec0e1213f6c3b24R572-R577 as an example

**Comment by tantaman** [comment]
If anyone else ends up here, I had to:

1. Create my data
2. Enqueue a micro task (e.g. `setImmediate`) to run `global.gc()`
3. Check that weak refs were cleaned up after said task completed

Any other way and the weak refs were not cleaned up.

e.g.,
```js
test('cleaned', async () => {
  const w = weak();
  await new Promise(resolve => setTimeout(resolve, 0));
  global.gc();
  expect(w.deref()).toEqual(null);
});

function weak() {
  return new WeakRef({});
}
```
